### PR TITLE
Refactor Transformer to include functionality about non-consuming inputs

### PR DESCRIPTION
**Pull request by JeppeBlixen**
I have updated the Transformer to incorporate the non-consuming behaviour from TransformerNC

The flag consume_inputs have been added with default true. This means the input dataset keys are consumed by default, and are therefore backwards kompatible.

I have updated the old TransformerNC to use this updated Transformer setting the consume_inputs flag to false. Therefore this change is also backwards kompatible with systems using TransformerNC.

I have added a deprecation warning to TransformerNC, at some point we should remove this transformer.

There are stille the work of updating alle the transformers that have been developed to stop using TransformerNC.

### CODE REVIEW

File: src/spetlr/etl/transformer.py [hunk 1]
```diff
@@ -1,29 +1,86 @@
...
+
 from pyspark.sql import DataFrame
 
 from .types import EtlBase, dataset_group
 
 
 class Transformer(EtlBase):
-    """If you only want to transform a single input dataframe,
+    """
+    This the base Transformer class for the OELT framework.
```
Comment (LauJohansson) [inline src/spetlr/etl/transformer.py:11]: This is*

Comment (LauJohansson) [inline src/spetlr/etl/transformer.py:11]: OETL*

File: src/spetlr/etl/transformer.py [hunk 2]
```diff
@@ -1,29 +1,86 @@
...
-            return {self.dataset_key: self.process(df)}
-        return {self.dataset_key: self.process_many(inputs)}
+        # If dataset_input_key_list is empty, it will be set to all keys in inputs
+        if len(self.dataset_input_key_list) == 0:
+            dataset_input_key_list = list(inputs.keys())
+        else:
+            dataset_input_key_list = self.dataset_input_key_list
+
+        # Call process or process_many depending on the input_key count
+        if len(dataset_input_key_list) == 1:
```
Comment (LauJohansson) [inline src/spetlr/etl/transformer.py:68]: just a discussion subject: Could we imagine a scenario, where the len(dataset_input_key_list) == 0 ? do **inputs* always have at least one key?

File: src/spetlr/etl/transformer.py [hunk 3]
```diff
@@ -1,29 +1,86 @@
...
+            df = self.process(inputs[dataset_input_key_list[0]])
+        else:
+            datasetFilteret = {
```
Comment (LauJohansson) [inline src/spetlr/etl/transformer.py:71]: maybe rename to "datasetFiltered" ?

File: src/spetlr/etl/transformer_nc.py [hunk 4]
```diff
@@ -22,36 +25,15 @@ def __init__(
...
-
-        inputs[self.dataset_output_key] = df
-        return inputs
+        super().__init__(
+            dataset_input_keys=dataset_input_keys,
+            dataset_output_key=dataset_output_key,
+            consume_inputs=False,
+        )
+
+        warnings.warn(
```
Comment (LauJohansson) [inline src/spetlr/etl/transformer_nc.py:34]: use deprecation libary instead

File: tests/cluster/etl/test_transformer.py [hunk 5]
```diff
@@ -14,19 +14,70 @@ def setUpClass(cls):
...
+        # Result contain single df
         result = self.transformer.etl({"df": self.df})
+        self.assertEqual(1, len(result))
         self.assertEqual({"TestTransformer"}, set(result.keys()))
-        self.assertIs(list(result.values())[0], self.df)
+        self.assertIs(self.df, result["TestTransformer"])
 
     def test_process_many(self):
+        # Process_many is called with multiple inputs
+        # Result contain union of input dfs og length 6
```
Comment (LauJohansson) [inline tests/cluster/etl/test_transformer.py:27]: and*

File: tests/cluster/etl/test_transformer.py [hunk 6]
```diff
@@ -14,19 +14,70 @@ def setUpClass(cls):
...
+        # Set dataset_input_keys to control what dfs in dataset to handle
+        # Process is called with single input
+        # Result contain two keys due to 'df2' not being consumed
+        # when not in dataset_input_keys
+        result = TestTransformer(dataset_input_keys="df1").etl(
+            {"df1": self.df, "df2": self.df}
+        )
+        self.assertEqual(2, len(result))
+        self.assertEqual({"TestTransformer", "df2"}, set(result.keys()))
+        self.assertIs(self.df, result["TestTransformer"])
```
Comment (LauJohansson) [inline tests/cluster/etl/test_transformer.py:59]: Should we add self.assertIs(self.df, result["df2"]) ?

File: tests/cluster/etl/test_transformer.py [hunk 7]
```diff
@@ -14,19 +14,70 @@ def setUpClass(cls):
...
+
+    def test_multiple_dataset_input_keys(self):
+        # Set dataset_input_keys to control what dfs in dataset to handle
+        # Process_many is called with multiple inputs
+        # Result contain two keys due to 'df3' not being consumed
```
Comment (LauJohansson) [inline tests/cluster/etl/test_transformer.py:64]: If df3 is not consumed, wouldnt it then contain THREE keys?

File: tests/cluster/etl/test_transformer.py [hunk 8]
```diff
@@ -14,19 +14,70 @@ def setUpClass(cls):
...
+        # Result contain the key 'my_key'
+        result = TestTransformer(dataset_output_key="my_key").etl({"df": self.df})
+        self.assertEqual(1, len(result))
+        self.assertEqual({"my_key"}, set(result.keys()))
+        self.assertIs(self.df, result["my_key"])
+
+    def test_single_dataset_input_key(self):
+        # Set dataset_input_keys to control what dfs in dataset to handle
+        # Process is called with single input
+        # Result contain two keys due to 'df2' not being consumed
```
Comment (LauJohansson) [inline tests/cluster/etl/test_transformer.py:52]: if the testtransformer is consume=true by default - then the result will have only 1 output, right?

File: tests/cluster/etl/test_transformer.py
Comment (LauJohansson) [inline tests/cluster/etl/test_transformer.py:1]: actually no need for this to be in the cluster folder (it can be runned locally). Maybe we should do a cleanup in another PR

File: src/spetlr/etl/transformer.py (hunk 2 above)
Comment (JeppeBlixen) [reply src/spetlr/etl/transformer.py:68]: That would mean you have to extractions and run a transformer on empty dataset dict.
In this case we could pass the inputs directly though, or we might thow an error.

File: src/spetlr/etl/transformer.py (hunk 3 above)
Comment (JeppeBlixen) [reply src/spetlr/etl/transformer.py:71]: Renamed it to inputs_filtered, makes more sense

File: tests/cluster/etl/test_transformer.py (hunk 7 above)
Comment (JeppeBlixen) [reply tests/cluster/etl/test_transformer.py:64]: 'df3' is not consumed, but 'df1' and 'df2' are merged into 'TestTransformer' therefore 2 keys.

File: src/spetlr/etl/transformer.py (hunk 2 above)
Comment (mrmasterplan) [reply src/spetlr/etl/transformer.py:68]: A transformer without inputs should be called an extractor for clarity, so yes, we can assume that len(inputs)>0

File: docs/etl/README.md [hunk 9]
```diff
@@ -621,13 +621,14 @@ etl = (
     .extract_from(OfficeEmployeeExtractor(dataset_key="df_employee"))
     .extract_from(OfficeBirthdaysExtractor(dataset_key="df_birthdays"))
     .transform_with(
-        IntegerTransformerNC(
+        IntegerTransformer(
             dataset_input_keys="df_employee",
```
Comment (mrmasterplan) [inline docs/etl/README.md:625]: should this be ["df_employee"] ?

File: docs/etl/README.template.md [hunk 10]
```diff
@@ -168,12 +168,15 @@ It is important that the first transformer is a `MultiInputTransformer` when hav
 ### Example-7
 
 This example illustrates the use of `TransformerNC`.
```
Comment (mrmasterplan) [inline docs/etl/README.template.md:170]: should this line go?

File: docs/etl/README.template.md [hunk 11]
```diff
@@ -168,12 +168,15 @@ It is important that the first transformer is a `MultiInputTransformer` when hav
...
+This example illustrates the use of `Transformer` in a non conuming context.
 The job here is to join the two extracted dataframes - an employees dataframe and a birthdays dataframe.
-But, before the birthdays can be joined onto the employees, the employees dataframe require a transformation step.
+But, before the birthdays can be join onto the employees, the employees dataframe require a transformation step.
```
Comment (mrmasterplan) [inline docs/etl/README.template.md:173]: can be joined is correct. Actually later in the line, "require" should be "requires"

File: docs/etl/README.template.md [hunk 12]
```diff
@@ -168,12 +168,15 @@ It is important that the first transformer is a `MultiInputTransformer` when hav
...
 As the transformation step of employees is handled by an `TransformerNC`, it does not consume the other inputs from the `dataset_group`.
```
Comment (mrmasterplan) [inline docs/etl/README.template.md:174]: should this line be gone?

File: docs/etl/README.template.md [hunk 13]
```diff
@@ -168,12 +168,15 @@ It is important that the first transformer is a `MultiInputTransformer` when hav
...
+As the transformation step of employees is handled by an non conuming `Transformer`, it does not consume the other inputs from the `dataset_group`.
 Hence, birthdays is still available from the inputs - even after the transformation of employees.
 Then both frames can be joined and the final dataframe saved via an `Loader`.
 When working with `TransformerNC` it is important to mind that dataset keys are crucial.
```
Comment (mrmasterplan) [inline docs/etl/README.template.md:178]: Im not sure this diff view is correct. This line should be gone.

File: docs/etl/README.md (hunk 9 above)
Comment (mrmasterplan) [reply docs/etl/README.md:625]: I see now that you actually allow both ways of specifying the input. I am a fan of the zen of python: `There should be one-- and preferably only one --obvious way to do it.` (https://peps.python.org/pep-0020/) but I will not insist on this. If you like it like this, then just resolve the conversation.

**Review by mrmasterplan — APPROVED** [review]
All comments are minor and relate to documentation and style. Approved.

File: docs/etl/README.md (hunk 9 above)
Comment (JeppeBlixen) [reply docs/etl/README.md:625]: I have decided to define dataset_input_keys only as List[str], and changed the rest accordingly.

**Review by MartinBoge — APPROVED** [review]
